R/join_keys.R

```diff
@@ -0,0 +1,620 @@
+    value <- original_value
+  }
+
+  join_keys_obj
+}
+
+#' @rdname join_keys
+#' @export
+#' @examples
+#'
+#' jk <- join_keys()
+#' jk[["ds1"]] <- list()
+#' jk[["ds2"]][["ds3"]] <- "key"
+#'
+#' jk <- join_keys()
+#' jk[["ds1"]] <- list()
+#' jk[["ds2"]][["ds3"]] <- "key"
+#' jk[["ds4"]] <- list(ds5 = "new")
+#'
+#' jk <- join_keys()
+#' jk[["ds2"]][["ds3"]] <- "key"
+#' jk[["ds2"]][["ds3"]] <- NULL
+#' jk
+`[[<-.join_keys` <- function(join_keys_obj, dataset_1, value) {
+  checkmate::assert_string(dataset_1)
+
+  # Accepting 1 subscript with valid `value` formal
+  checkmate::assert_list(value, names = "named", types = "character", null.ok = TRUE)
+
+  # Normalize values
+  norm_value <- lapply(names(value), function(.x) {
+    get_keys(join_key(dataset_1, .x, value[[.x]]))
+  })
+
+  names(norm_value) <- names(value)
+  value <- norm_value
+
+  # Keep original parameters as variables will be overwritten for `NextMethod` call
+  original_value <- value
+  ds1 <- dataset_1
+
+  # In case an pair is removed, also remove the symmetric pair
+  removed_names <- setdiff(names(join_keys_obj[[dataset_1]]), names(value))
+  if (length(removed_names) > 0) {
+    for (.x in removed_names) {
+      value <- join_keys_obj[[.x]]
+      value[[ds1]] <- NULL
+      dataset_1 <- .x
+      join_keys_obj <- NextMethod("[[<-", join_keys_obj)
+    }
+
+    # Restore original values
+    dataset_1 <- ds1
+    value <- original_value
+  }
+
+  join_keys_obj <- NextMethod("[[<-", join_keys_obj)
+
+  # Iterate on all new values to create symmetrical pair
+  for (ds2 in names(original_value)) {
+    if (ds2 == ds1) next
+
+    value <- join_keys_obj[[ds2]] %||% list()
+    new_value <- original_value[[ds2]]
+
+    if (checkmate::test_character(new_value, min.len = 1, names = "unnamed")) {
+      new_value <- setNames(new_value, new_value)
+    } else if (
+      checkmate::test_character(new_value, min.len = 1)
+    ) {
+      # Invert key
+      new_value <- setNames(names(new_value), new_value)
+    }
+
+    # Change variables for NextMethod call
+    dataset_1 <- ds2
+    value[[ds1]] <- new_value
+    join_keys_obj <- NextMethod("[[<-", join_keys_obj)
+  }
+
+  join_keys_obj
+}
+
+#' @rdname merge_join_keys
+#' @keywords internal
+merge_join_keys <- function(join_keys_obj, new_join_keys) {
+  UseMethod("merge_join_keys", join_keys_obj)
+}
+
+#' @rdname merge_join_keys
+#' @keywords internal
+merge_join_keys.default <- function(join_keys_obj, new_join_keys) {
+  merge_join_keys(join_keys(join_keys_obj), new_join_keys)
+}
+
+#' Merging a list (or one) of `join_keys` objects into the current `join_keys` object
+#'
+#' @rdname merge_join_keys
+#'
+#' @param join_keys_obj (`join_keys`) object to merge the new_join_keys.
+#' @param new_join_keys  `list` of `join_keys` objects or single `join_keys` object
+#'
+#' @return a new `join_keys` object with the resulting merge.
+#'
+#' @keywords internal
+merge_join_keys.join_keys <- function(join_keys_obj, new_join_keys) {
```

Comment:
There is no generic for `merge_join_keys`, so `merge_join_keys.join_keys` is just a normal function (not S3 dispatched in generic function). It means that `.join_keys` is redundant and you need to add assert. I like the fact that it is not S3 so it doesn't pollute namespace.
  
```suggestion
merge_join_keys <- function(join_keys_obj, new_join_keys) {
  checkmate::assert_class(join_keys_obj, "join_keys")
```